lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -511,20 +598,92 @@ LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
+    // inductionReg by the step of the For loop. Gets an adder by mimicking what
+    // is dones in `buildLibraryOp`. There may be a more concise way to do this.
+    Type regWidth = inductionReg.getOut().getType();
+    // Adder should have same width as the inductionReg.
+    SmallVector<Type> types(3, regWidth);
+    auto addOp = getState<ComponentLoweringState>()
+                     .getNewLibraryOpInstance<calyx::AddLibOp>(
+                         rewriter, forOp.getLoc(), types);
+
+    auto directions = addOp.portDirections();
+    // For an add operation, we expect two input ports and one output port
+    SmallVector<Value, 2> opInputPorts;
+    Value opOutputPort;
+    for (auto dir : enumerate(directions)) {
+      if (dir.value() == calyx::Direction::Input)
+        opInputPorts.push_back(addOp.getResult(dir.index()));
+      else {
+        if (opOutputPort != nullptr) {
+          assert(false &&
+                 "Expected Calyx add operation to have only one output port");
+        }
+        opOutputPort = addOp.getResult(dir.index());
+      }
+    }
+    assert(opInputPorts.size() == 2 &&
+           "Expected an add Calyx operation to have 2 inputs");
+
+    // "Latch Group" increments inductionReg by forLoop's step value.
+    calyx::ComponentOp componentOp =
+        getState<ComponentLoweringState>().getComponentOp();
+    auto groupOp = calyx::createGroup<calyx::GroupOp>(
+        rewriter, componentOp, forOp.getLoc(),
+        "incr_" + getState<ComponentLoweringState>().getUniqueName(forOp) +
+            "_induction_var");
+    rewriter.setInsertionPointToEnd(groupOp.getBodyBlock());
+
+    // Assign inductionReg.out to the left port of the adder.
+    auto leftOp = opInputPorts.front();
+    rewriter.create<calyx::AssignOp>(forOp.getLoc(), leftOp,
+                                     inductionReg.getOut());
+    // Assign forOp.getConstantStep to the right port of the adder.
+    auto rightOp = opInputPorts.back();
+    rewriter.create<calyx::AssignOp>(
+        forOp.getLoc(), rightOp,
+        createConstant(forOp->getLoc(), rewriter, componentOp,
+                       regWidth.getIntOrFloatBitWidth(),
+                       forOp.getConstantStep().value().getSExtValue()));
+    // Assign adder's output port to inductionReg.
+    buildAssignmentsForRegisterWrite(rewriter, groupOp, componentOp,
+                                     inductionReg, opOutputPort);
+    // Set group as For Loop's "latch" group.
+    getState<ComponentLoweringState>().setForLoopLatchGroup(forOpInterface,
+                                                            groupOp);
+    getState<ComponentLoweringState>().registerEvaluatingGroup(opOutputPort,
+                                                               groupOp);
     return success();
+  }
+  // If yieldOp for a for loop is not empty, then we do not transform for loop.
+  if (dyn_cast<scf::ForOp>(yieldOp->getParentOp())) {
+    yieldOp.getOperation()->emitError()
```

Comment:
Can we add an error test for these?

Reply:
Did this, and also added a test case for the error for the case where we don't know the bound.